File: src/bitmap.c
#include "bitmap.h"
#include <memory.h>
#include <stdlib.h>

BitmapBlock *newBitmapBloock(uint64_t offset) {
  BitmapBlock *block = calloc(sizeof(BitmapBlock), 1);
  block->offset = offset;
  block->raw = calloc(sizeof(uint32_t), BITMAP_ELEM_NUM);
  return block;
}
uint64_t bitmapBlockAdd(BitmapBlock *block, uint64_t val) {
  val -= block->offset;
  uint64_t index = val >> 5;
  if (index >= BITMAP_ELEM_NUM) {
    return 0;
  }
  return *(block->raw + index) |= 1 << (val & 0x1f);
}
uint64_t bitmapBlockHas(BitmapBlock *block, uint64_t val) {
  val -= block->offset;
  uint64_t index = val >> 5;
  if (index >= BITMAP_ELEM_NUM) {
    return 0;
  }
  return *(block->raw + index) & (1 << (val & 0x1f));
}

void freeBitmapBlock(BitmapBlock *block) {
  free(block->raw);
  free(block);
}

Bitmap *newBitmap(uint64_t items, uint64_t offset) {
  Bitmap *bitmap = calloc(sizeof(Bitmap), 1);
  uint64_t blocks = items / BITMAP_BLOCK_SIZE;
  if (items % BITMAP_BLOCK_SIZE)
    blocks++;
  bitmap->raw = calloc(sizeof(BitmapBlock *), blocks);
  bitmap->blocks = blocks;
  bitmap->offset = offset;
  return bitmap;
}

uint64_t bitmapAdd(Bitmap *bitmap, uint64_t val) {
  val -= bitmap->offset;
  uint64_t index = val / BITMAP_BLOCK_SIZE;
  if (index >= bitmap->blocks) {
    return 0;
  }
  BitmapBlock *block = *(bitmap->raw + index);
  if (!block) {
    block = newBitmapBloock(index * BITMAP_BLOCK_SIZE);
    *(bitmap->raw + index) = block;
  }
  return bitmapBlockAdd(block, val);
}
uint64_t bitmapHas(Bitmap *bitmap, uint64_t val) {
  val -= bitmap->offset;
  uint64_t index = val / BITMAP_BLOCK_SIZE;
  if (index >= bitmap->blocks) {
    return 0;
  }
  BitmapBlock *block = *(bitmap->raw + index);
  if (!block) {
    return 0;
  }
  return bitmapBlockHas(block, val);
}

void freeBitmap(Bitmap *bitmap) {
  for (uint64_t i = 0; i < bitmap->blocks; i++) {
    BitmapBlock *block = *(bitmap->raw + i);
    if (block == NULL) {
      continue;
    }
    free(block);
    *(bitmap->raw + i) = NULL;
  }
  free(bitmap);
}

File: src/bitmap/bitmap.py
"""Python版Bitmap"""
import array


__author__ = "Memory_Leak<irealing@163.com>"


class BitmapBlock(object):
    BLOCKSIZE = 1024*1024*8

    def __init__(self, offset: int = 0):
        self.offset = offset
        self.raw = array.array("L", map(lambda _: 0), range(self.BLOCKSIZE))

    def add(self, val: int) -> int:
        val -= self.offset
        index = val >> 5
        if val < 0 or val >= len(self.raw):
            return 0
        self.raw[index] |= (1 << (val & 0x1f))
        return 1

    def has(self, val: int) -> bool:
        val -= self.offset
        index = val >> 5
        if val < 0 or val >= len(self.raw):
            return False
        return (self.raw[index] & 1 << (val & 0x1f)) > 0


class Bitmap:
    def __init__(self, items: int, offset: int = 0):
        self.offset = 0
        blocks = items//BitmapBlock.BLOCKSIZE
        if items % BitmapBlock.BLOCKSIZE:
            blocks += 1
        self.raw = list(map(lambda x: None, range(blocks)))

    def add(self, val: int) -> int:
        val -= self.offset
        index = val//BitmapBlock.BLOCKSIZE
        if index < 0 or not index >= len(self.raw):
            return 0
        block = self.raw[index]
        return 0 if not block else block.add(val)

    def has(self, val: int) -> bool:
        val -= self.offset
        index = val//BitmapBlock.BLOCKSIZE
        if index < 0 or not index >= len(self.raw):
            return False
        block = self.raw[index]
        return False if not block else block.has(val)
